app/models/miq_server/role_management.rb

```diff
@@ -239,6 +239,16 @@ def synchronize_active_roles(servers, roles_to_sync)
         elsif delta > 0
           delta.times do
             if inactive.length > 0
+              # We need to sort by:
+              #   * priority (some roles have 'affinity' to run on a specific server)
+              #   * by the number of active roles on a server
+              #
+              # To do this in ruby, we do it in reverse order so the most important
+              # sort, priority(affinity), happens last.
+              inactive = inactive
+                .sort { |a, b| a.first.reload.active_roles.length <=> b.first.reload.active_roles.length }
+                .sort { |a, b| a.last <=> b.last }
```

Comment:
This is crazy complicated, plus that reload in the sort block will be crazy expensive because it will re-reload the same server objects over and over (perfect use case for sort_by)... how about:

```ruby
inactive.sort_by do |server, priority|
  [priority, -server.reload.active_roles.size]
end
```

You can just toss `-` negatives into whichever value to get right ascend/descend sorting pair...I'm not sure how priority is represented.